From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ROCKNIX <rocknix@rocknix.org>
Date: Mon, 17 Feb 2026 00:00:00 +0100
Subject: [PATCH] net: rtl8733bu: use dev_addr_set() for kernel 6.x compat

Since Linux 5.17, writing directly to netdev->dev_addr via memcpy is
forbidden. The kernel validates the address through dev_addr_check()
and triggers a WARNING when the address doesn't match the internal
dev_addr_list, which also prevents cfg80211/ConnMan from recognizing
the WiFi interface.

Replace all _rtw_memcpy(ndev->dev_addr, ...) calls with dev_addr_set()
which properly updates both dev_addr and the address list.

Fixes: "WARNING: CPU: ... at net/core/dev_addr_lists.c:520 dev_addr_check"
Fixes: "connmanctl enable wifi" returning "wifi is not available"

Signed-off-by: ROCKNIX <rocknix@rocknix.org>
---
 os_dep/linux/mlme_linux.c | 2 +-
 os_dep/linux/os_intfs.c   | 8 ++++----
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/os_dep/linux/mlme_linux.c b/os_dep/linux/mlme_linux.c
index 1111111..2222222 100644
--- a/os_dep/linux/mlme_linux.c
+++ b/os_dep/linux/mlme_linux.c
@@ -412,7 +412,7 @@ void rtw_os_ndev_register_mgnt_frame(struct net_device *ndev)
 	mac[4] = 0x11;
 	mac[5] = 0x12;
 
-	_rtw_memcpy(pnetdev->dev_addr, mac, ETH_ALEN);
+	dev_addr_set(pnetdev, mac);
 
 
 	rtw_netif_carrier_off(pnetdev);
diff --git a/os_dep/linux/os_intfs.c b/os_dep/linux/os_intfs.c
index 3333333..4444444 100644
--- a/os_dep/linux/os_intfs.c
+++ b/os_dep/linux/os_intfs.c
@@ -1639,7 +1639,7 @@ static int rtw_net_set_mac_address(struct net_device *pnetdev, void *addr)
 	}
 
 	_rtw_memcpy(adapter_mac_addr(padapter), sa->sa_data, ETH_ALEN); /* set mac addr to adapter */
-	_rtw_memcpy(pnetdev->dev_addr, sa->sa_data, ETH_ALEN); /* set mac addr to net_device */
+	dev_addr_set(pnetdev, sa->sa_data); /* set mac addr to net_device */
 
 #if 0
 	if (rtw_is_hw_init_completed(padapter)) {
@@ -2197,7 +2197,7 @@ int rtw_os_ndev_register(_adapter *adapter, const char *name, bool rtnl_lock_nee
 	/* alloc netdev name */
 	rtw_init_netdev_name(ndev, name);
 
-	_rtw_memcpy(ndev->dev_addr, adapter_mac_addr(adapter), ETH_ALEN);
+	dev_addr_set(ndev, adapter_mac_addr(adapter));
 
 	/* Tell the network stack we exist */
 
@@ -3269,7 +3269,7 @@ static int _netdev_open(struct net_device *pnetdev)
 		rtw_mbid_camid_alloc(padapter, adapter_mac_addr(padapter));
 #endif
 		rtw_init_wifidirect_addrs(padapter, adapter_mac_addr(padapter), adapter_mac_addr(padapter));
-		_rtw_memcpy(pnetdev->dev_addr, adapter_mac_addr(padapter), ETH_ALEN);
+		dev_addr_set(pnetdev, adapter_mac_addr(padapter));
 	}
 #endif /*CONFIG_PLATFORM_INTEL_BYT*/
 
@@ -4025,7 +4025,7 @@ int _netdev_vir_if_open(struct net_device *pnetdev)
 		rtw_mbid_camid_alloc(padapter, adapter_mac_addr(padapter));
 #endif
 		rtw_init_wifidirect_addrs(padapter, adapter_mac_addr(padapter), adapter_mac_addr(padapter));
-		_rtw_memcpy(pnetdev->dev_addr, adapter_mac_addr(padapter), ETH_ALEN);
+		dev_addr_set(pnetdev, adapter_mac_addr(padapter));
 #endif /* CONFIG_PLATFORM_INTEL_BYT */
 
 		rtw_clr_surprise_removed(padapter);
